### PR TITLE
fix(bedrock): sanitize replayed tool names before converse

### LINKED ISSUE #4585: Bedrock: hallucinated tool names with invalid characters crash subsequent API calls
## Description

When using `BedrockConverseModel`, if the model hallucinates a tool name containing characters outside `[a-zA-Z0-9_-]` (e.g., a dot or space), PydanticAI records the `ToolCallPart` in the message history. On the next API round-trip, PydanticAI serializes this history back to Bedrock, which rejects the entire request with:

```
ValidationException: 1 validation error detected: Value at 'messages.N.member.content.M.member.toolUse.name'
failed to satisfy constraint: Member must satisfy regular expression pattern: [a-zA-Z0-9_-]+
```

This is unrecoverable — the invalid tool name is baked into the history, so every subsequent request in the same agent run will fail.

## Root cause

In `pydantic_ai/models/bedrock.py`, `_map_tool_call` passes the tool name through without validation:

```python
@staticmethod
def _map_tool_call(t: ToolCallPart) -> ContentBlockOutputTypeDef:
    return {
        'toolUse': {'toolUseId': _utils.guard_tool_call_id(t=t), 'name': t.tool_name, 'input': t.args_as_dict()}
    }
```

## Expected behavior

PydanticAI should sanitize or reject tool names that don't match Bedrock's `[a-zA-Z0-9_-]+` constraint before sending them to the API. This could be done either:

1. When recording the `ToolCallPart` from the model's response (replace invalid chars)
2. When serializing history in `_map_tool_call` (sanitize on output)
3. Both

## Reproduction

1. Create an agent with `BedrockConverseModel` and several tools
2. Use a system prompt that encourages the model to call tools with dotted names (e.g., "use module.function_name syntax")
3. When the model hallucinates a tool name like `search.evidence`, the first retry works, but the second API call crashes because the hallucinated name is in the history

## Workaround

We added a `history_processor` that sanitizes tool names before each API call:

```python
import re
from dataclasses import replace

_BEDROCK_TOOL_NAME_RE = re.compile(r"[^a-zA-Z0-9_-]")

def sanitize_tool_names(messages):
    result = []
    for msg in messages:
        if isinstance(msg, ModelResponse):
            new_parts = []
            modified = False
            for part in msg.parts:
                if isinstance(part, ToolCallPart) and _BEDROCK_TOOL_NAME_RE.search(part.tool_name):
                    new_parts.append(replace(part, tool_name=_BEDROCK_TOOL_NAME_RE.sub("_", part.tool_name)))
                    modified = True
                else:
                    new_parts.append(part)
            result.append(replace(msg, parts=new_parts) if modified else msg)
        else:
            result.append(msg)
    return result
```

## Environment

- pydantic-ai-slim 1.63.0
- Bedrock model: `us.anthropic.claude-sonnet-4-5-20250929-v1:0`
- Python 3.12

### PR DESCRIPTION
**Pull request by pandego**
## Summary
Fix Bedrock message-history replay failures when tool names contain characters outside Bedrock's allowed pattern.

This change sanitizes replayed `ToolCallPart.tool_name` values in `BedrockConverseModel._map_tool_call` so outbound `toolUse.name` always matches `[a-zA-Z0-9_-]+`.

## Changes
- add a Bedrock-specific sanitizer for tool names used during message serialization
- replace invalid characters with `_`
- map empty names to `_` so serialized names always satisfy Bedrock's `+` requirement
- add focused tests for invalid-character and empty-name cases

## Test plan
- `uv run ruff check pydantic_ai_slim/pydantic_ai/models/bedrock.py tests/models/test_bedrock.py`
- `uv run pytest tests/models/test_bedrock.py -k 'map_tool_call_sanitizes_invalid_tool_name or map_tool_call_replaces_empty_tool_name'`

Closes #4585

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/models/bedrock.py [hunk 1]
```diff
@@ -954,7 +962,11 @@ async def _map_user_prompt(  # noqa: C901
     @staticmethod
     def _map_tool_call(t: ToolCallPart) -> ContentBlockOutputTypeDef:
         return {
-            'toolUse': {'toolUseId': _utils.guard_tool_call_id(t=t), 'name': t.tool_name, 'input': t.args_as_dict()}
+            'toolUse': {
+                'toolUseId': _utils.guard_tool_call_id(t=t),
+                'name': _sanitize_bedrock_tool_name(t.tool_name),
+                'input': t.args_as_dict(),
+            }
         }
```
[the remark this answers is not in the document]
Comment (pandego) [reply pydantic_ai_slim/pydantic_ai/models/bedrock.py:994]: Thanks - fixed in e523b8f0. I now sanitize tool definition names in `_map_tool_definition`, keep a sanitized->original map, and restore original names in both `_process_response` and streaming (`_get_event_iterator`) before constructing `ToolCallPart`.

File: tests/models/test_bedrock.py [hunk 2]
```diff
@@ -1786,6 +1786,34 @@ async def test_bedrock_no_tool_choice(bedrock_provider: BedrockProvider):
...
+
+
+def test_bedrock_map_tool_call_replaces_empty_tool_name(bedrock_provider: BedrockProvider):
+    model = BedrockConverseModel('us.anthropic.claude-sonnet-4-5-20250929-v1:0', provider=bedrock_provider)
+
+    tool_use = model._map_tool_call(  # pyright: ignore[reportPrivateUsage]
+        ToolCallPart(tool_name='', args={'q': 'x'}, tool_call_id='tooluse_123')
+    )
+
+    assert tool_use['toolUse']['name'] == '_'
```
[the remark this answers is not in the document]
Comment (pandego) [reply tests/models/test_bedrock.py:1815]: Good callout. I kept these as focused unit tests for the Bedrock name-normalization helpers (plus added mapping/collision unit coverage) to keep this regression deterministic and fast. Existing integration coverage for Bedrock request/response paths remains in place.

File: pydantic_ai_slim/pydantic_ai/models/bedrock.py [hunk 3]
```diff
@@ -86,6 +87,13 @@
 _SUPPORTED_IMAGE_FORMATS = ('jpeg', 'png', 'gif', 'webp')
 _SUPPORTED_VIDEO_FORMATS = ('mkv', 'mov', 'mp4', 'webm', 'flv', 'mpeg', 'mpg', 'wmv', 'three_gp')
 _SUPPORTED_DOCUMENT_FORMATS = ('pdf', 'txt', 'csv', 'doc', 'docx', 'xls', 'xlsx', 'html', 'md')
+_BEDROCK_TOOL_NAME_INVALID_CHAR_RE = re.compile(r'[^a-zA-Z0-9_-]')
+
+
+def _sanitize_bedrock_tool_name(name: str) -> str:
+    """Normalize tool names to match Bedrock's required ``[a-zA-Z0-9_-]+`` pattern."""
+    sanitized = _BEDROCK_TOOL_NAME_INVALID_CHAR_RE.sub('_', name)
+    return sanitized or '_'
```
[the remark this answers is not in the document]
Comment (pandego) [reply pydantic_ai_slim/pydantic_ai/models/bedrock.py:96]: Implemented in e523b8f0: added bidirectional mapping via `_sanitized_tool_name_to_original_name`, and restore on both non-stream and stream paths before creating `ToolCallPart`.

[the remark this answers is not in the document]
Comment (pandego) [reply pydantic_ai_slim/pydantic_ai/models/bedrock.py:96]: Implemented in e523b8f0: added collision detection in `_get_tools` and now raise `UserError` when two original tool names sanitize to the same Bedrock name. Added test coverage for this case.

File: pydantic_ai_slim/pydantic_ai/models/bedrock.py [hunk 4]
```diff
@@ -389,11 +398,28 @@ def supported_builtin_tools(cls) -> frozenset[type[AbstractBuiltinTool]]:
         return frozenset({CodeExecutionTool})
 
     def _get_tools(self, model_request_parameters: ModelRequestParameters) -> list[ToolTypeDef]:
-        return [self._map_tool_definition(r) for r in model_request_parameters.tool_defs.values()]
+        self._sanitized_tool_name_to_original_name = {}
```
[the remark this answers is not in the document]
Comment (pandego) [reply pydantic_ai_slim/pydantic_ai/models/bedrock.py]: Great catch - fixed in dba50d81. I removed per-instance mutable mapping state and made tool-name mapping request-local (`_build_sanitized_tool_name_map`), then thread that mapping through `request`/`request_stream` into `_process_response`/`BedrockStreamedResponse`.

File: pydantic_ai_slim/pydantic_ai/models/bedrock.py [hunk 5]
```diff
@@ -1130,7 +1168,7 @@ async def _get_event_iterator(self) -> AsyncIterator[ModelResponseStreamEvent]:
                         tool_use = delta['toolUse']
                         maybe_event = self._parts_manager.handle_tool_call_delta(
                             vendor_part_id=index,
-                            tool_name=tool_use.get('name'),
+                            tool_name=self._restore_tool_name(tool_use.get('name', '')),
```
[the remark this answers is not in the document]
Comment (pandego) [reply pydantic_ai_slim/pydantic_ai/models/bedrock.py]: Fixed in 0eeb6222. Streaming deltas now pass `tool_name=None` when the delta omits `name`, preserving existing `ToolCallPartDelta` semantics and snapshots.

File: tests/models/test_bedrock.py [hunk 6]
```diff
@@ -1786,6 +1786,65 @@ async def test_bedrock_no_tool_choice(bedrock_provider: BedrockProvider):
...
+
+
+def test_bedrock_tool_name_sanitization_collision_raises_user_error(bedrock_provider: BedrockProvider):
+    model = BedrockConverseModel('us.anthropic.claude-sonnet-4-5-20250929-v1:0', provider=bedrock_provider)
+    tool_a = ToolDefinition(name='search.foo', parameters_json_schema={'type': 'object'})
+    tool_b = ToolDefinition(name='search:foo', parameters_json_schema={'type': 'object'})
+    mrp = ModelRequestParameters(function_tools=[tool_a, tool_b])
+
+    with pytest.raises(UserError, match='sanitization collision'):
+        model._get_tools(mrp)  # pyright: ignore[reportPrivateUsage]
```
[the remark this answers is not in the document]
Comment (pandego) [reply tests/models/test_bedrock.py:1848]: Acknowledged. I kept these as focused unit tests because this file already validates many Bedrock transforms via private helpers. Integration coverage remains in existing Bedrock VCR tests, and I also validated affected streaming paths locally.

[the remark this answers is not in the document]
Comment (pandego) [reply tests/models/test_bedrock.py:1848]: Good point. For this PR I kept deterministic unit coverage for the sanitization/mapping edge-cases and validated behavior through existing Bedrock integration stream tests. If maintainers want, we can add a dedicated VCR scenario in a follow-up.

**Comment by DouweM** [comment]
Closing in favor of https://github.com/pydantic/pydantic-ai/pull/4713
